DOCS: rewording the raft Handling Failures and recovery procedure sections (#13177)
scylladb/scylladb: docs/architecture/raft.rst

Handling Failures
------------------
Raft requires a quorum of nodes in a cluster to proceed. As long as the majority of nodes are alive and connected, schema and topology updates proceed unaffected.
When the node that was down is up again, it first contacts the cluster to fetch the latest schema and then starts serving queries.

> ScyllaDB data path, read and write requests, works with :term:`Consistency Level (CL)` , and are not dependent on Raft quorum.

Author: ```suggestion
ScyllaDB data path read and write requests work with :term:Consistency Level (CL) and is not dependent on Raft quorum.
```

Reviewer: Either "work"+"are" (both plural, referring to "read and write requests"), or "works"+"is" (both singular, referring to "data path"). Here you combined "work" (plural) with "is" (singular) which doesn't make sense.

Raft requires a quorum of nodes in a cluster to proceed. As long as the majority of nodes are alive and connected, schema and topology updates proceed unaffected.
When the node that was down is up again, it first contacts the cluster to fetch the latest schema and then starts serving queries.

> ScyllaDB data path read and write requests work with :term:Consistency Level (CL) and is not dependent on Raft quorum.

Reviewer 2: either works and is or work and are. either data path serves as an adjective, or as a noun/subject.

     - Schema updates are possible and safe.
     - Try restarting the node. If the node is dead, :doc:`replace it with a new node </operating-scylla/procedures/cluster-management/replace-dead-node/>`.
   * - 2 nodes

>      - Data is available for reads and writes, but schema and toplogy changes are only possible once the quorum is restored.

Author: Aha! this is more complex than the original version and my suggestion, and desrves its own section.
Is this list will be true for future release as well? (at least 5.3?)

>      - Data is available for reads and writes, but schema and toplogy changes are only possible once the quorum is restored.

Reviewer: The original version only mentioned that schema changes can be done only with quorum and that is correct.
The new version proposes that topology changes would be possible with only quorum which is not correct.
So I would say the original version is fine.

Topology changes did not change from older versions. They always required all nodes to be alive (except `removenode` with `--ignore-dead-nodes`). The existing documentation covers this I think, so there's no reason to write more about it.

>  Is this list will be true for future release as well? (at least 5.3?)

Probably yes, I think only tablets will be able to change that.

   * - 2 DCs
     - Data is available for reads and writes, but schema and toplogy changes are only possible once the quorum is restored.
     - When the DCs come back online, restart the nodes. If at least one DC fails to come back online and the nodes are lost, consult the :ref:`Raft recovery section <recover-raft-procedure>`.

.. _recover-raft-procedure:

Raft Recovery Procedure
=======================

.. versionadded:: 5.2

Prerequisite
~~~~~~~~~~~~
                  
The Raft recovery procedure applies to the following situations:

* A majority of nodes failed and are **irrecoverable**. Raft can only progress if a majority (quorum) of nodes are available. If it is possible to restart your nodes, do that instead of the recovery procedure.
* :ref:`The internal Raft upgrade procedure <verify-raft-procedure>` got stuck because one of your nodes failed in the middle of the procedure and is irrecoverable,

Make sure that the dead nodes are irrecoverable, and are not, for example, temporarily partitioned away due to a network failure. If "dead" nodes return to life, they might communicate and interfere with the recovery procedure.

> It's good practice to ensure that using firewall rules or otherwise isolates your live nodes and blocks communication from these "dead" nodes.

Author: @kostja @asias  is this still relevant?
https://github.com/scylladb/scylladb/commit/12ab2c3d8d6925a14749be61b90a3032c994f1f8 Prevent removed node to restart and join the cluster

> It's good practice to ensure that using firewall rules or otherwise isolates your live nodes and blocks communication from these "dead" nodes.

Author: grammarly suggest otherwise!

Reviewer: The statement is saying that you should use firewall rules, or **you** should isolate and block communication.

Reviewer: The original statement was:
>    If you have no means of ensuring that these irrecoverable nodes won't come back to life and communicate with the rest of the cluster, setup firewall rules or otherwise isolate your alive nodes to reject any communication attempts from these dead nodes.

It's saying:
setup firewall rules
OR
using some other method (that's what the "otherwise" word means here), isolate your alive nodes to reject any communication attempts...

the general idea is: you need to isolate the alive nodes from the unavailable node if you're not sure that it's truly dead. You can do that by setting up firewall rules or otherwise. (i.e., using a different method of your choice.)

So it's B

Procedure
~~~~~~~~~

During the recovery procedure, you'll enter a special ``RECOVERY`` mode, remove all faulty nodes (using the standard :doc:`node removal procedure </operating-scylla/procedures/cluster-management/remove-node/>`), delete the internal Raft data, and execute a rolling restart. This will cause the cluster to perform the internal Raft upgrade procedure again, initializing the Raft algorithm from scratch. The recovery procedure applies to clusters not running Raft in the past and then had Raft enabled and to clusters bootstrapped with Raft.

.. note::

>    Entering ``RECOVERY`` mode requires a node restart. Restarting an additional node while some nodes are already dead may lead to unavailability data queries. For example, if you're using the standard RF=3, CL=QUORUM setup, and you're recovering from a stuck of upgrade procedure because one of your nodes is dead, restarting another node will cause temporary data query unavailability (until the node finishes restarting). Prepare your service for downtime before proceeding.

Author: restoring the wording as @kbr-scylla suggsted
